src/directions.js

```diff
@@ -132,46 +164,69 @@ var Directions = L.Class.extend({
         return this.getOrigin() && this.getDestination();
     },
 
-    query: function () {
+    query: function (opts) {
+        if (!opts) opts = {};
         if (!this.queryable()) return this;
 
         if (this._query) {
             this._query.abort();
         }
 
-        this._query = corslite(this.queryURL(), L.bind(function (err, resp) {
-            this._query = null;
+        if (this._geocode) {
+            this._geocode.abort();
+        }
 
-            if (err && err.type === 'abort') {
-                return;
-            }
+        var q = queue();
 
-            resp = resp || err;
+        var pts = [this.origin, this.destination].concat(this._waypoints);
+        for (var i in pts) {
+            if (!pts[i].geometry.coordinates) q.defer(L.bind(this.geocode, this), pts[i], opts.proximity || opts);
+        }
 
-            if (resp && resp.responseText) {
-                try {
-                    resp = JSON.parse(resp.responseText);
-                } catch (e) {
-                    resp = {error: resp.responseText};
+        q.await(L.bind(function(err, res) {
```

Comment:
They should fire an error event, because a) it's unlikely fallback to the original inputs will work (unless the user typed in comma separated decimal lng lat, the directions API will just error) and b) the user should get feedback in the error control about what went wrong.